Critical: reference search doesn't work for fields in JDK classes

Find reference for System.out after open java.lang.System class in vscode java. No references are shown

But take com.google.common.collect.MultimapBuilder and search for any field or constant. The reference works.

This issue cannot be reproduced in upstream JDT project.

I couldn't find the download file and tring building locally, But the ls build fails

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-p2-director-plugin:3.0.1:materialize-products (create-distributions) on project org.eclipse.jdt.ls.product: Installation of product languageServer.product for environment linux/gtk/x86_64 failed: Call to p2 director application failed with exit code 13. Program arguments were: [-metadataRepository
```

May be something to do with target file updates.

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ReferencesHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ReferencesHandler.java
@@ -21,12 +21,15 @@ import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.jdt.core.IAnnotation;
 import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.core.IClasspathContainer;
+import org.eclipse.jdt.core.IClasspathEntry;
 import org.eclipse.jdt.core.ICompilationUnit;
 import org.eclipse.jdt.core.IField;
 import org.eclipse.jdt.core.IJavaElement;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.core.IMemberValuePair;
 import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
 import org.eclipse.jdt.core.IType;
 import org.eclipse.jdt.core.ITypeRoot;
 import org.eclipse.jdt.core.JavaCore;
@@ -53,13 +56,13 @@ public final class ReferencesHandler {
 		this.preferenceManager = preferenceManager;
 	}
 
-	private IJavaSearchScope createSearchScope() throws JavaModelException {
+	private IJavaSearchScope createSearchScope(IJavaElement elementToSearch) throws JavaModelException {
 		IJavaProject[] projects = JavaCore.create(ResourcesPlugin.getWorkspace().getRoot()).getJavaProjects();
-		int scope = IJavaSearchScope.SOURCES;
-		if (preferenceManager.isClientSupportsClassFileContent()) {
-			scope |= IJavaSearchScope.APPLICATION_LIBRARIES;
+		int includeMask = IJavaSearchScope.SOURCES | IJavaSearchScope.REFERENCED_PROJECTS | IJavaSearchScope.APPLICATION_LIBRARIES;
+		if (isInsideJRE(elementToSearch)) {
+			includeMask |= IJavaSearchScope.SYSTEM_LIBRARIES;
 		}
-		return SearchEngine.createJavaSearchScope(projects, scope);
+		return SearchEngine.createJavaSearchScope(projects, includeMask);
 	}
 
 	public List<Location> findReferences(ReferenceParams param, IProgressMonitor monitor) {
@@ -148,7 +151,26 @@ public final class ReferencesHandler {
 		return declaringType.getFullyQualifiedName() + "." + declaringType.getElementName() + "Builder";
 	}
 
-	private void search(IJavaElement elementToSearch, final List<Location> locations, IProgressMonitor monitor, boolean isIncludeDeclaration) throws CoreException, JavaModelException {
+	// copied from org.eclipse.jdt.internal.ui.search.JavaSearchScopeFactory.isInsideJRE(IJavaElement)
+	private boolean isInsideJRE(IJavaElement element) {
+		IPackageFragmentRoot root = (IPackageFragmentRoot) element.getAncestor(IJavaElement.PACKAGE_FRAGMENT_ROOT);
+		if (root != null) {
+			try {
+				IClasspathEntry entry = root.getRawClasspathEntry();
+				if (entry.getEntryKind() == IClasspathEntry.CPE_CONTAINER) {
+					IClasspathContainer container = JavaCore.getClasspathContainer(entry.getPath(), root.getJavaProject());
+					return container != null && container.getKind() == IClasspathContainer.K_DEFAULT_SYSTEM;
+				}
+				return false;
+			} catch (JavaModelException e) {
+				JavaLanguageServerPlugin.logException(e);
+			}
+		}
+		return true; // include JRE in doubt
+	}
+
+	// for test purpose only
+	public void search(IJavaElement elementToSearch, final List<Location> locations, IProgressMonitor monitor, boolean isIncludeDeclaration) throws CoreException, JavaModelException {
 		boolean includeClassFiles = preferenceManager.isClientSupportsClassFileContent();
 		boolean includeDecompiledSources = preferenceManager.getPreferences().isIncludeDecompiledSources();
 		SearchEngine engine = new SearchEngine();
@@ -157,7 +179,7 @@ public final class ReferencesHandler {
 			SearchPattern patternDecl = SearchPattern.createPattern(elementToSearch, IJavaSearchConstants.DECLARATIONS);
 			pattern = SearchPattern.createOrPattern(pattern, patternDecl);
 		}
-		engine.search(pattern, new SearchParticipant[] { SearchEngine.getDefaultSearchParticipant() }, createSearchScope(), new SearchRequestor() {
+		engine.search(pattern, new SearchParticipant[] { SearchEngine.getDefaultSearchParticipant() }, createSearchScope(elementToSearch), new SearchRequestor() {
 
 			@Override
 			public void acceptSearchMatch(SearchMatch match) throws CoreException {

--- a/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/ReferencesHandlerTest.java
+++ b/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/ReferencesHandlerTest.java
@@ -19,10 +19,12 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
 import java.net.URI;
+import java.util.ArrayList;
 import java.util.List;
 
 import org.eclipse.core.resources.IProject;
 import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.core.IField;
 import org.eclipse.jdt.core.IJavaElement;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.core.IType;
@@ -173,6 +175,21 @@ public class ReferencesHandlerTest extends AbstractProjectsManagerBasedTest{
 		assertEquals(new Range(new Position(12, 20), new Position(12, 32)), l.getRange());
 		l = references.get(1);
 		assertEquals(new Range(new Position(14, 15), new Position(14, 25)), l.getRange());
+	}
+
+	// https://github.com/eclipse/eclipse.jdt.ls/issues/2405
+	@Test
+	public void testReferencesInJRE() throws Exception {
+		when(preferenceManager.isClientSupportsClassFileContent()).thenReturn(true);
+		IJavaProject javaProject = JavaCore.create(project);
+		IType system = javaProject.findType("java.lang.System");
+		IField field = system.getField("out");
+		assertTrue(field.exists());
+		List<Location> references = new ArrayList<>();
+		handler.search(field, references, monitor, true);
+		assertNotNull("findReferences should not return null", references);
+		Location location = references.stream().filter(r -> r.getUri().startsWith("jdt://contents/rtstubs.jar/java.lang/System.class")).findFirst().get();
+		assertNotNull(location);
 	}
 
 }

